include/stdexec/ranges.hpp

```diff
@@ -0,0 +1,244 @@
+          noexcept(__nothrow_constructible_from<
+                   __operation<_Self, _Receiver>,
+                   __copy_cvref_t<_Self, _Sender>,
+                   _Receiver&&,
+                   __copy_cvref_t<_Self, _Closure>>) -> __operation<_Self, _Receiver> {
+          return {
+            ((_Self &&) __self).__sndr_, (_Receiver &&) __rcvr, ((_Self &&) __self).__closure_};
+        }
+
+        template <__decays_to<__t> _Self, class _Env>
+        friend auto tag_invoke(get_completion_signatures_t, _Self&&, _Env)
+          -> dependent_completion_signatures<_Env>;
+
+        template <__decays_to<__t> _Self, class _Env>
+        friend auto tag_invoke(get_completion_signatures_t, _Self&&, _Env)
+          -> __completion_signatures<_Self, _Env>
+          requires true;
+
+        friend auto tag_invoke(get_env_t, const __t& __self) //
+          noexcept(__nothrow_callable<get_env_t, const _Sender&>)
+            -> __call_result_t<get_env_t, const _Sender&> {
+          return get_env(__self.__sndr_);
+        }
+      };
+    };
+
+    template <class _Callable_error, template <class> class __closure, class DerivedClosure>
+    struct __range_adaptor {
+      template <class _Sender, class _Fun>
+      using __sender =
+        __t<__closure_adaptor::
+              __sender<stdexec::__id<__decay_t<_Sender>>, __closure<_Fun>, _Callable_error>>;
+
+      template <sender _Sender, __movable_value _Fun>
+        requires(!__tag_invocable_with_completion_scheduler<
+                  DerivedClosure,
+                  set_value_t,
+                  _Sender,
+                  _Fun>)
+             && (!tag_invocable<DerivedClosure, _Sender, _Fun>) && sender<__sender<_Sender, _Fun>>
+      __sender<_Sender, _Fun> operator()(_Sender&& __sndr, _Fun __fun) const {
+        return __sender<_Sender, _Fun>{(_Sender &&) __sndr, __closure<_Fun>((_Fun &&) __fun)};
+      }
+
+      template <sender _Sender, __movable_value _Fun>
+        requires __tag_invocable_with_completion_scheduler<DerivedClosure, set_value_t, _Sender, _Fun>
+      sender auto operator()(_Sender&& __sndr, _Fun __fun) const
+        noexcept(nothrow_tag_invocable<
+                 DerivedClosure,
+                 __completion_scheduler_for<_Sender, set_value_t>,
+                 _Sender,
+                 _Fun>) {
+        auto __sched = get_completion_scheduler<set_value_t>(get_env(__sndr));
+        return tag_invoke(
+          DerivedClosure{},
+          std::move(__sched),
+          (_Sender &&) __sndr,
+          __closure<_Fun>((_Fun &&) __fun));
+      }
+
+      template <sender _Sender, __movable_value _Fun>
+        requires(!__tag_invocable_with_completion_scheduler<
+                  DerivedClosure,
+                  set_value_t,
+                  _Sender,
+                  _Fun>)
+             && tag_invocable<DerivedClosure, _Sender, _Fun>
+      sender auto operator()(_Sender&& __sndr, _Fun __fun) const
```

Comment:
I think this can be further generalized. 

We should accept anything that can be passed to the respective `std::view` and returns not a viewable_range, e.g. we would accept `transform(fun)` but not `transform(range, fun)`